File: link_checker.py


### 1. `link_checker.py` (Python)

```python
# link_checker.py — Python версия

import sys
import argparse
import json
import csv
import time
import requests
from concurrent.futures import ThreadPoolExecutor, as_completed
from urllib.parse import urljoin, urlparse
from colorama import init, Fore, Style

init(autoreset=True)

class LinkChecker:
    def __init__(self, timeout=10, max_redirects=10, headers=None):
        self.timeout = timeout
        self.max_redirects = max_redirects
        self.headers = headers or {}
        self.session = requests.Session()
        self.session.headers.update(self.headers)

    def check_url(self, url):
        """Проверяет один URL, отслеживая редиректы."""
        start_time = time.time()
        result = {
            'url': url,
            'final_url': None,
            'status_code': None,
            'redirects': [],
            'time': 0,
            'size': 0,
            'error': None
        }

        current_url = url
        redirect_count = 0

        try:
            while redirect_count <= self.max_redirects:
                response = self.session.get(current_url, timeout=self.timeout, allow_redirects=False, stream=True)
                status = response.status_code
                result['redirects'].append({
                    'url': current_url,
                    'status_code': status,
                    'headers': dict(response.headers)
                })

                if 300 <= status < 400:
                    location = response.headers.get('Location')
                    if not location:
                        break
                    # Обработка относительных редиректов
                    if not urlparse(location).netloc:
                        location = urljoin(current_url, location)
                    current_url = location
                    redirect_count += 1
                    if redirect_count > self.max_redirects:
                        result['error'] = f'Слишком много редиректов (> {self.max_redirects})'
                        break
                else:
                    result['status_code'] = status
                    result['final_url'] = current_url
                    result['size'] = len(response.content) if response.content else 0
                    break
            else:
                result['error'] = f'Превышен лимит редиректов ({self.max_redirects})'

        except requests.exceptions.RequestException as e:
            result['error'] = str(e)
        except Exception as e:
            result['error'] = f'Неизвестная ошибка: {e}'

        result['time'] = time.time() - start_time
        return result

    def check_multiple(self, urls, workers=10):
        """Проверяет несколько URL параллельно."""
        results = {}
        with ThreadPoolExecutor(max_workers=workers) as executor:
            future_to_url = {executor.submit(self.check_url, url): url for url in urls}
            for future in as_completed(future_to_url):
                url = future_to_url[future]
                try:
                    results[url] = future.result()
                except Exception as e:
                    results[url] = {'error': str(e)}
        return results

    def print_results(self, results):
        """Выводит результаты в терминал с цветами."""
        for url, data in results.items():
            print(f"\n{Fore.CYAN}🔍 Проверка: {url}{Style.RESET_ALL}")
            if data.get('error'):
                print(f"{Fore.RED}❌ Ошибка: {data['error']}{Style.RESET_ALL}")
                continue

            status = data.get('status_code')
            final_url = data.get('final_url', url)
            time_taken = data.get('time', 0)
            size = data.get('size', 0)
            redirects = data.get('redirects', [])

            if status and 200 <= status < 300:
                status_color = Fore.GREEN
                status_text = f"{status} OK"
            elif status and 300 <= status < 400:
                status_color = Fore.YELLOW
                status_text = f"{status} Redirect"
            else:
                status_color = Fore.RED
                status_text = f"{status} Error" if status else "Error"

            print(f"{status_color}✅ {final_url}{Style.RESET_ALL}")
            print(f"   Статус: {status_color}{status_text}{Style.RESET_ALL}")
            print(f"   Время: {time_taken:.2f} сек")
            print(f"   Размер: {size} байт")
            if redirects:
                print(f"   Редиректы: {len(redirects)}")
                for i, r in enumerate(redirects, 1):
                    print(f"     {i}. {r['url']} -> {r['status_code']}")

    def save_json(self, results, filename):
        with open(filename, 'w', encoding='utf-8') as f:
            json.dump(results, f, indent=2, ensure_ascii=False)
        print(f"{Fore.GREEN}💾 Сохранено JSON: {filename}{Style.RESET_ALL}")

    def save_csv(self, results, filename):
        import csv
        with open(filename, 'w', newline='', encoding='utf-8') as f:
            writer = csv.writer(f)
            writer.writerow(['URL', 'Final URL', 'Status', 'Time (s)', 'Size (bytes)', 'Redirects', 'Error'])
            for url, data in results.items():
                writer.writerow([
                    url,
                    data.get('final_url', ''),
                    data.get('status_code', ''),
                    f"{data.get('time', 0):.3f}",
                    data.get('size', 0),
                    len(data.get('redirects', [])),
                    data.get('error', '')
                ])
        print(f"{Fore.GREEN}💾 Сохранено CSV: {filename}{Style.RESET_ALL}")

def main():
    parser = argparse.ArgumentParser(description='Link Checker (Redirects)')
    parser.add_argument('urls', nargs='*', help='Список URL для проверки')
    parser.add_argument('--file', '-f', help='Файл со списком URL (по одному на строку)')
    parser.add_argument('--timeout', type=int, default=10, help='Таймаут (сек)')
    parser.add_argument('--max-redirects', type=int, default=10, help='Максимум редиректов')
    parser.add_argument('--output-json', help='Сохранить в JSON')
    parser.add_argument('--output-csv', help='Сохранить в CSV')
    parser.add_argument('--workers', type=int, default=10, help='Количество потоков')
    parser.add_argument('--header', '-H', action='append', help='Добавить заголовок (key: value)')
    args = parser.parse_args()

    print(f"{Fore.CYAN}🔗 Link Checker (Python){Style.RESET_ALL}")

    # Сбор URL
    urls = []
    if args.urls:
        urls.extend(args.urls)
    if args.file:
        with open(args.file, 'r', encoding='utf-8') as f:
            for line in f:
                line = line.strip()
                if line:
                    urls.append(line)
    if not urls:
        print("❌ Нет URL для проверки.")
        sys.exit(1)

    # Заголовки
    headers = {}
    if args.header:
        for h in args.header:
            if ':' in h:
                key, value = h.split(':', 1)
                headers[key.strip()] = value.strip()

    checker = LinkChecker(timeout=args.timeout, max_redirects=args.max_redirects, headers=headers)
    results = checker.check_multiple(urls, workers=args.workers)
    checker.print_results(results)

    if args.output_json:
        checker.save_json(results, args.output_json)
    if args.output_csv:
        checker.save_csv(results, args.output_csv)

if __name__ == "__main__":
    main()

File: link_checker.go
// link_checker.go — Go версия

package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"flag"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"strings"
	"sync"
	"time"
)

type RedirectStep struct {
	URL        string            `json:"url"`
	StatusCode int               `json:"status_code"`
	Headers    map[string]string `json:"headers"`
}

type Result struct {
	URL        string           `json:"url"`
	FinalURL   string           `json:"final_url"`
	StatusCode int              `json:"status_code"`
	Redirects  []RedirectStep   `json:"redirects"`
	Time       float64          `json:"time"`
	Size       int64            `json:"size"`
	Error      string           `json:"error,omitempty"`
}

type LinkChecker struct {
	Timeout      time.Duration
	MaxRedirects int
	Headers      map[string]string
	Client       *http.Client
}

func NewLinkChecker(timeout int, maxRedirects int, headers map[string]string) *LinkChecker {
	return &LinkChecker{
		Timeout:      time.Duration(timeout) * time.Second,
		MaxRedirects: maxRedirects,
		Headers:      headers,
		Client: &http.Client{
			Timeout: time.Duration(timeout) * time.Second,
			CheckRedirect: func(req *http.Request, via []*http.Request) error {
				if len(via) >= maxRedirects {
					return fmt.Errorf("слишком много редиректов")
				}
				return nil
			},
		},
	}
}

func (lc *LinkChecker) checkURL(urlStr string) Result {
	start := time.Now()
	result := Result{URL: urlStr, Redirects: []RedirectStep{}}

	req, err := http.NewRequest("GET", urlStr, nil)
	if err != nil {
		result.Error = err.Error()
		return result
	}

	for k, v := range lc.Headers {
		req.Header.Set(k, v)
	}

	resp, err := lc.Client.Do(req)
	if err != nil {
		result.Error = err.Error()
		return result
	}
	defer resp.Body.Close()

	// Собираем редиректы из истории
	history := []RedirectStep{}
	currentURL := urlStr
	for _, r := range resp.Request.Response {
		// Неполно, но мы можем отследить вручную
	}
	// Простейший вариант: используем resp.Request.URL как конечный
	finalURL := resp.Request.URL.String()
	if resp.StatusCode >= 300 && resp.StatusCode < 400 {
		// Попытка получить Location
		loc := resp.Header.Get("Location")
		if loc != "" {
			// Добавляем шаг редиректа
			history = append(history, RedirectStep{
				URL:        currentURL,
				StatusCode: resp.StatusCode,
				Headers:    map[string]string{"Location": loc},
			})
			// Разрешаем относительный URL
			if !strings.HasPrefix(loc, "http") {
				base, _ := url.Parse(currentURL)
				rel, _ := url.Parse(loc)
				loc = base.ResolveReference(rel).String()
			}
			// Рекурсивно проверяем следующий URL
			nextResult := lc.checkURL(loc)
			result.FinalURL = nextResult.FinalURL
			result.StatusCode = nextResult.StatusCode
			result.Redirects = append(history, nextResult.Redirects...)
			result.Size = nextResult.Size
			result.Time = time.Since(start).Seconds()
			return result
		}
	}

	// Нет редиректа или финальный ответ
	result.FinalURL = finalURL
	result.StatusCode = resp.StatusCode
	body, _ := io.ReadAll(resp.Body)
	result.Size = int64(len(body))
	result.Time = time.Since(start).Seconds()
	return result
}

func (lc *LinkChecker) checkMultiple(urls []string, workers int) map[string]Result {
	results := make(map[string]Result)
	var mu sync.Mutex
	var wg sync.WaitGroup
	sem := make(chan struct{}, workers)

	for _, u := range urls {
		wg.Add(1)
		go func(url string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()
			res := lc.checkURL(url)
			mu.Lock()
			results[url] = res
			mu.Unlock()
		}(u)
	}
	wg.Wait()
	return results
}

func printResults(results map[string]Result) {
	for url, res := range results {
		fmt.Printf("\n\x1b[36m🔍 Проверка: %s\x1b[0m\n", url)
		if res.Error != "" {
			fmt.Printf("\x1b[31m❌ Ошибка: %s\x1b[0m\n", res.Error)
			continue
		}
		statusColor := "\x1b[32m"
		statusText := fmt.Sprintf("%d OK", res.StatusCode)
		if res.StatusCode >= 300 && res.StatusCode < 400 {
			statusColor = "\x1b[33m"
			statusText = fmt.Sprintf("%d Redirect", res.StatusCode)
		} else if res.StatusCode >= 400 {
			statusColor = "\x1b[31m"
			statusText = fmt.Sprintf("%d Error", res.StatusCode)
		}
		fmt.Printf("%s✅ %s\x1b[0m\n", statusColor, res.FinalURL)
		fmt.Printf("   Статус: %s%s\x1b[0m\n", statusColor, statusText)
		fmt.Printf("   Время: %.2f сек\n", res.Time)
		fmt.Printf("   Размер: %d байт\n", res.Size)
		if len(res.Redirects) > 0 {
			fmt.Printf("   Редиректы: %d\n", len(res.Redirects))
			for i, r := range res.Redirects {
				fmt.Printf("     %d. %s -> %d\n", i+1, r.URL, r.StatusCode)
			}
		}
	}
}

func saveJSON(results map[string]Result, filename string) error {
	data, err := json.MarshalIndent(results, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, data, 0644)
}

func saveCSV(results map[string]Result, filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()
	writer.Write([]string{"URL", "Final URL", "Status", "Time (s)", "Size (bytes)", "Redirects", "Error"})
	for u, r := range results {
		writer.Write([]string{
			u,
			r.FinalURL,
			fmt.Sprintf("%d", r.StatusCode),
			fmt.Sprintf("%.3f", r.Time),
			fmt.Sprintf("%d", r.Size),
			fmt.Sprintf("%d", len(r.Redirects)),
			r.Error,
		})
	}
	return nil
}

func main() {
	var urlsStr string
	var filePath string
	var timeout int
	var maxRedirects int
	var outputJSON string
	var outputCSV string
	var workers int
	var headersStr string

	flag.StringVar(&urlsStr, "urls", "", "Список URL через запятую")
	flag.StringVar(&filePath, "file", "", "Файл со списком URL")
	flag.IntVar(&timeout, "timeout", 10, "Таймаут (сек)")
	flag.IntVar(&maxRedirects, "max-redirects", 10, "Максимум редиректов")
	flag.StringVar(&outputJSON, "output-json", "", "Сохранить в JSON")
	flag.StringVar(&outputCSV, "output-csv", "", "Сохранить в CSV")
	flag.IntVar(&workers, "workers", 10, "Количество потоков")
	flag.StringVar(&headersStr, "header", "", "Заголовки (key1:value1,key2:value2)")
	flag.Parse()

	fmt.Println("\x1b[36m🔗 Link Checker (Go)\x1b[0m")

	var urls []string
	if urlsStr != "" {
		urls = append(urls, strings.Split(urlsStr, ",")...)
	}
	if filePath != "" {
		file, err := os.Open(filePath)
		if err != nil {
			fmt.Printf("❌ Ошибка открытия файла: %v\n", err)
			os.Exit(1)
		}
		defer file.Close()
		scanner := bufio.NewScanner(file)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line != "" {
				urls = append(urls, line)
			}
		}
	}
	if len(urls) == 0 {
		fmt.Println("❌ Нет URL для проверки.")
		os.Exit(1)
	}

	headers := map[string]string{}
	if headersStr != "" {
		for _, h := range strings.Split(headersStr, ",") {
			if parts := strings.SplitN(h, ":", 2); len(parts) == 2 {
				headers[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
			}
		}
	}

	checker := NewLinkChecker(timeout, maxRedirects, headers)
	results := checker.checkMultiple(urls, workers)
	printResults(results)

	if outputJSON != "" {
		if err := saveJSON(results, outputJSON); err == nil {
			fmt.Printf("\x1b[32m💾 Сохранено JSON: %s\x1b[0m\n", outputJSON)
		}
	}
	if outputCSV != "" {
		if err := saveCSV(results, outputCSV); err == nil {
			fmt.Printf("\x1b[32m💾 Сохранено CSV: %s\x1b[0m\n", outputCSV)
		}
	}
}
